### PR TITLE
os/bluestore: move assert of read/write to base class

### PR DESCRIPTION
**Pull request by mychoxin**
Signed-off-by: mychoxin <mychoxin@gmail.com>

### CODE REVIEW

**Comment by joscollin** [comment]
Jenkins retest this please

File: src/os/bluestore/PMEMDevice.cc [hunk 1]
```diff
@@ -217,9 +217,7 @@ int PMEMDevice::write(uint64_t off, bufferlist& bl, bool buffered)
 {
   uint64_t len = bl.length();
   dout(20) << __func__ << " " << off << "~" << len  << dendl;
-  assert(len > 0);
-  assert(off < size);
-  assert(off + len <= size);
+  assert(is_valid_io(off, len));
```
Comment (joscollin) [inline src/os/bluestore/PMEMDevice.cc]: It is different here. Only 3 assert statements

File: src/os/bluestore/PMEMDevice.cc [hunk 2]
```diff
@@ -261,9 +259,7 @@ int PMEMDevice::read(uint64_t off, uint64_t len, bufferlist *pbl,
 		      bool buffered)
 {
   dout(5) << __func__ << " " << off << "~" << len  << dendl;
-  assert(len > 0);
-  assert(off < size);
-  assert(off + len <= size);
+  assert(is_valid_io(off, len));
```
Comment (joscollin) [inline src/os/bluestore/PMEMDevice.cc]: ditto

File: src/os/bluestore/PMEMDevice.cc [hunk 3]
```diff
@@ -286,9 +282,8 @@ int PMEMDevice::aio_read(uint64_t off, uint64_t len, bufferlist *pbl,
 
 int PMEMDevice::read_random(uint64_t off, uint64_t len, char *buf, bool buffered)
 {
-  assert(len > 0);
-  assert(off < size);
-  assert(off + len <= size);
+  dout(5) << __func__ << " " << off << "~" << len << dendl;
+  assert(is_valid_io(off, len));
```
Comment (joscollin) [inline src/os/bluestore/PMEMDevice.cc]: ditto

File: src/os/bluestore/PMEMDevice.cc (hunk 1 above)
Comment (mychoxin) [inline src/os/bluestore/PMEMDevice.cc]: @majianpeng For this place, I think the original code misses 2 assert, do you agree?

**Comment by mychoxin** [comment]
@majianpeng

File: src/os/bluestore/PMEMDevice.cc (hunk 1 above)
Comment (majianpeng) [inline src/os/bluestore/PMEMDevice.cc]: Yes, PMEMDevice is based on byte address.  So offset && length are not limited by block_size (in fact by directIO requirement)

**Comment by mychoxin** [comment]
@joscollin I modified it, please take a look.

File: src/os/bluestore/BlockDevice.h [hunk 4]
```diff
@@ -148,6 +148,13 @@ class BlockDevice {
   virtual int invalidate_cache(uint64_t off, uint64_t len) = 0;
   virtual int open(const std::string& path) = 0;
   virtual void close() = 0;
+  virtual bool is_valid_io(uint64_t off, uint64_t len) {
```
Comment (tchaikov) [inline src/os/bluestore/BlockDevice.h]: no need to make this a virtual method. a static method would suffice.

Comment (joscollin) [inline src/os/bluestore/BlockDevice.h]: Right. `is_valid_io` is not getting called via an interface to be `virtual`. A simple overriding in `PMEMDevice` works well here.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-08-20_01:49:41-rados-wip-kefu-testing-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-08-19_02:32:27-rados-wip-kefu-testing-distro-basic-mira/

**Comment by mychoxin** [comment]
@tchaikov @joscollin Inoder to access the class member variable, I just remove 'virtual' not change it to static function, what do you think of it?

File: src/os/bluestore/BlockDevice.h [hunk 5]
```diff
@@ -148,6 +148,13 @@ class BlockDevice {
   virtual int invalidate_cache(uint64_t off, uint64_t len) = 0;
   virtual int open(const std::string& path) = 0;
   virtual void close() = 0;
+  bool is_valid_io(uint64_t off, uint64_t len) {
```
Comment (tchaikov) [inline src/os/bluestore/BlockDevice.h]: mark this method `private` and `const`. and you can parenthesizes the condition expressions  like
```c
return (off % block_size == 0 &&
        len % block_size == 0 && 
 ...
        off + len <= size);
```

File: src/os/bluestore/PMEMDevice.h [hunk 6]
```diff
@@ -60,6 +60,11 @@ class PMEMDevice : public BlockDevice {
   int invalidate_cache(uint64_t off, uint64_t len) override;
   int open(const std::string &path) override;
   void close() override;
+  bool is_valid_io(uint64_t off, uint64_t len) {
```
Comment (tchaikov) [inline src/os/bluestore/PMEMDevice.h]: the same applies.

File: src/os/bluestore/BlockDevice.h (hunk 5 above)
Comment (mychoxin) [inline src/os/bluestore/BlockDevice.h]: done, BlockDevice::is_valid_io is called by derived class, so I put it under protected. please have a look.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-08-24_05:32:54-rados-wip-kefu-testing-distro-basic-smithi/

**Comment by joscollin** [comment]
@mychoxin: Sorry, I was on leave. Your changes LGTM.

**Pull request merged**